corehq/ex-submodules/casexml/apps/case/xform.py

```diff
@@ -336,25 +336,22 @@ def _get_or_update_cases(xforms, case_db):
     """
     # have to apply the deprecations before the updates
     sorted_forms = sorted(xforms, key=lambda f: 0 if is_deprecation(f) else 1)
+    touched_cases = {}
     for xform in sorted_forms:
         for case_update in get_case_updates(xform):
             case_doc = _get_or_update_model(case_update, xform, case_db)
+            touched_cases[case_doc['_id']] = case_doc
             if case_doc:
                 # todo: legacy behavior, should remove after new case processing
                 # is fully enabled.
                 if xform._id not in case_doc.xform_ids:
                     case_doc.xform_ids.append(xform.get_id)
-                case_db.set(case_doc.case_id, case_doc)
             else:
                 logging.error(
                     "XForm %s had a case block that wasn't able to create a case! "
                     "This usually means it had a missing ID" % xform.get_id
                 )
 
-    # at this point we know which cases we want to update so copy this away
-    # this prevents indices that end up in the cache from being added to the return value
-    touched_cases = copy.copy(case_db.cache)
```

Comment:
I think [this logic here](https://github.com/dimagi/commcare-hq/blob/master/corehq/ex-submodules/casexml/apps/case/util.py#L129) is the main place where this can go wrong and result in unexpected things.